spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -193,47 +193,438 @@ For example, calling an `@Outgoing` annotated method from user code will not pos
+| This method is called once; at assembly time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+ProcessorBuilder<Message<I>, Message<O>> method();
+----
+| Returns a `ProcessorBuilder` consuming incoming `Message` instances and produces `Message` instances.
+| This method is called once; at assembly time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+ProcessorBuilder<I, O> method();
+----
+| Returns a Reactive Streams processor that consuming incoming _payload_ instances and produces _payload_ instances.
+| This method is called once; at assembly time.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Publisher<Message<O>> method(Message<I> msg)
+----
+| Returns a Reactive Streams `Publisher` for each incoming `Message`.
+The returned `Publisher` can be empty or emits multiple `Message` instances.
+If the returned `Publisher` emits several elements, these elements are _flatten_ in the outgoing stream as a concatenation of elements.
+| This method is called for every incoming message. Never concurrently, so the stream returned previously must complete before having the method invoked with the next incoming `Message`.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+Publisher<O> method(I payload)
+----
+| Returns a Reactive Streams `Publisher` for each incoming _payload_.
+The returned `Publisher` can be empty or emits multiple _payload_ instances.
+If the returned `Publisher` emits several elements, these elements are _flatten_ in the outgoing stream as a concatenation of elements.
+The Reactive Messaging implementation must create new `Message` instances for each emitted _payloads_ as well as extracing the payload for each incoming `Message`.
+| This method is called for every incoming message. Never concurrently, so the stream returned previously must complete before having the method invoked with the next incoming `Message`.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+PublisherBuilder<Message<O>> method(Message<I> msg)
+----
+| Returns a `PublisherBuilder` for each incoming `Message`.
+The returned `PublisherBuilder` can be empty or emits multiple `Message` instances.
+If the returned `PublisherBuilder` emits several elements, these elements are _flatten_ in the outgoing stream as a concatenation of elements.
+| This method is called for every incoming message. Never concurrently, so the stream returned previously must complete before having the method invoked with the next incoming `Message`.
+
+|
+[source,java]
+----
+@Incoming("in")
+@Outgoing("out") 
+PublisherBuilder<O> method(I payload)
+----
+| Returns a `PublisherBuilder` for each incoming _payload_.
+The returned `PublisherBuilder` can be empty or emits multiple _payload_ instances.
+If the returned `PublisherBuilder` emits several elements, these elements are _flatten_ in the outgoing stream as a concatenation of elements.
```

Comment:
If the stream emitted from the built Publisher...
Also the exact meaning of "_flatten_" is not clear to me - what is used to do the flattening?